### PR TITLE
RHAISTRAT-1014, RHAIENG-2609: chore(dependencies): update LLM Compressor to 0.9.0

### PR DESCRIPTION
**Pull request by jiridanek**
https://issues.redhat.com/browse/RHAIENG-2609
https://issues.redhat.com/browse/RHAISTRAT-1014

## Description

This commit updates the LLM Compressor library to version 0.9.0 across workbench and runtime environments.

* https://github.com/opendatahub-io/notebooks/pull/2795

It introduces a new meta-package to centralize related dependencies, creating a single source of truth for easier management. Several core libraries, including Numpy, Transformers, and Accelerate, were also upgraded to newer versions to maintain compatibility.

> [x] RHOAI 3.3 Pipeline images are built with the exact same version of llm-compressor and compressed-tensors to ensure serialization compatibility.

Environment lock files and image manifests were synchronized to reflect these updates, and the dependency locking process was refined to handle the new package structure.

## How Has This Been Tested?

Self checklist (all need to be checked):
- [x] Ensure that you have run `make test` (`gmake` on macOS) before asking for review
- [x] Changes to everything except `Dockerfile.konflux` files should be done in `odh/notebooks` and automatically synced to `rhds/notebooks`. For Konflux-specific changes, modify `Dockerfile.konflux` files directly in `rhds/notebooks` as these require special attention in the downstream repository and flow to the upcoming RHOAI release.

## Merge criteria:

- [x] The commits are squashed in a cohesive manner and have meaningful messages.
- [ ] Testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).
- [x] The developer has manually tested the changes and verified that the changes work

## Summary by CodeRabbit

* **New Features**
  * Centralized meta-dependency introduced to manage LLM Compressor-related packages.

* **Updates**
  * LLM-Compressor upgraded to 0.9.
  * NumPy standardized to 2.3.5 across environments.
  * Lockfile generation updated to exclude emitting the meta-dependency package.

* **Tests**
  * Validation updated to recognize and skip meta-dependency packages during lockfile checks.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

### CODE REVIEW

**Comment by jiridanek** [comment]
@coderabbitai review

File: dependencies/odh-notebooks-meta-llmcompressor-deps/pyproject.toml [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+    "pillow==11.3.0",
+    "auto-round==0.9.2",
+    "compressed-tensors==0.13.0",
+
+    # see https://redhat-internal.slack.com/archives/C079FE5H94J/p1756933987117709?thread_ts=1756933154.294109&cid=C079FE5H94J
+    # for a discussion on what version of torch to use
+
+    # we will ignore their torch preference for now
+    #"torch==2.9.1",
+
```
[the remark this answers is not in the document]
Comment (jiridanek) [reply dependencies/odh-notebooks-meta-llmcompressor-deps/pyproject.toml]: Do we? It's not mentioned in https://issues.redhat.com/browse/RHAISTRAT-1014?focusedId=28766926&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-28766926

File: dependencies/odh-notebooks-meta-llmcompressor-deps/pyproject.toml [hunk 2]
```diff
@@ -0,0 +1,38 @@
+[project]
```
[the remark this answers is not in the document]
Comment (jiridanek) [reply dependencies/odh-notebooks-meta-llmcompressor-deps/pyproject.toml]: No, I'm open to discuss any other suggestion to satisfy the requirement from https://issues.redhat.com/browse/RHAISTRAT-1014

> Version Drift: If the Workbench image lags behind the runtime image (or vice versa), models compressed in one environment may fail to load in the other due to serialization format mismatches in compressed-tensors.

[the remark this answers is not in the document]
Comment (jiridanek) [reply dependencies/odh-notebooks-meta-llmcompressor-deps/pyproject.toml]: add directory inside pytorch llm compressor? but which one, workbench or runtime?

File: scripts/pylocks_generator.sh [hunk 3]
```diff
@@ -194,6 +194,7 @@ for TARGET_DIR in "${TARGET_DIRS[@]}"; do
       --universal \
       --no-annotate \
       --quiet \
+      --no-emit-package odh-notebooks-meta-llmcompressor-deps \
```
Comment (atheo89) [inline scripts/pylocks_generator.sh]: How it will look like this command if we do the same for the rest of the images? 🤔

File: runtimes/rocm-pytorch/ubi9-python-3.12/pyproject.toml [hunk 4]
```diff
@@ -14,7 +14,7 @@ dependencies = [
     "boto3~=1.40.52",
     "kafka-python-ng~=2.2.3",
     "matplotlib~=3.10.7",
-    "numpy~=2.3.4",
+    "numpy~=2.3.5",
```
[the remark this answers is not in the document]
Comment (jiridanek) [reply runtimes/rocm-pytorch/ubi9-python-3.12/pyproject.toml]: we already used 2.3 before and it seemed to work, and uv resolves this

**Comment by jiridanek** [comment]
About that masking thing

# Gemini says

 This is a **version compatibility issue** between `transformers` 4.54.0 and the Qwen2 model's attention masking mechanism during GPTQ calibration. The error occurs because the `padding_mask` tensor has an unexpected 1D shape when the code expects 2D (batch, sequence).

### Root Cause
The nested `vmap` operations in `transformers/masking_utils.py` are failing when indexing `padding_mask[batch_idx, kv_idx]`. This is likely a bug in transformers 4.54.0's SDPA (Scaled Dot Product Attention) mask creation that was fixed in later versions, but you're constrained by llmcompressor compatibility.

### Recommended Solutions

#### 1. **Use Sequential Calibration Pipeline** (Quick Fix)
The warning explicitly recommends this. Modify your GPTQ recipe/config:

```python
# In your quantization config/recipe
recipe = """
quant_stage:
  quant_modifiers:
    GPTQModifier:
      config_groups:
        group_0:
          weights:
            num_bits: 4
            type: int
            symmetric: true
          targets: ["Linear"]
      # Add this line
      calibration_pipeline: "sequential"  # Instead of "basic"
"""
```

#### 2. **Disable SDPA Mechanism**
Force the model to use the older attention implementation:

```python
import torch
from transformers import AutoModelForCausalLM

# Load model with SDP disabled
model = AutoModelForCausalLM.from_pretrained(
    "Qwen/Qwen2-1.5B",
    torch_dtype=torch.float16,
    attn_implementation="eager"  # or "flash_attention_2" if available
)
```

#### 3. **Upgrade Transformers Strategically**
Try a newer version that fixes this but might still work with llmcompressor:

```bash
pip install transformers==4.56.1 tokenizers==0.21.1
```

This version is newer but still close enough that compatibility may hold. **If llmcompressor complains**, you may need to pin an older compatible version of llmcompressor too.

#### 4. **Modify Calibration Data Collator**
The issue could be how `neuralmagic/calibration` is formatted. Try preprocessing it explicitly:

```python
from datasets import load_dataset
from transformers import AutoTokenizer

tokenizer = AutoTokenizer.from_pretrained("Qwen/Qwen2-1.5B")

def preprocess(examples):
    return tokenizer(examples["text"], padding="max_length", max_length=1024, truncation=True)

calib_dataset = load_dataset("neuralmagic/calibration", split="train[:64]")
calib_dataset = calib_dataset.map(preprocess, batched=True, remove_columns=calib_dataset.column_names)
calib_dataset.set_format(type="torch", columns=["input_ids", "attention_mask"])
```

#### 5. **Downgrade PyTorch (Last Resort)**
The `vmap` behavior changed in recent PyTorch versions. If using PyTorch 2.5+, try:

```bash
pip install torch==2.4.1 torchvision==0.19.1 --index-url https://download.pytorch.org/whl/rocm6.4
```

### Immediate Action
Try **Solution #1 first** as it's the most targeted fix and aligns with the library's own recommendation. If that fails, **Solution #2** is the next best bet since it bypasses the problematic SDPA masking entirely.

Which approach would you like to try? If you share your GPTQ script/config, I can provide more specific code modifications.

**Comment by jiridanek** [comment]
# Kimi says

 This is a **version compatibility issue** between `transformers` 4.54.0 and the Qwen2 model's attention masking mechanism during GPTQ calibration. The error occurs because the `padding_mask` tensor has an unexpected 1D shape when the code expects 2D (batch, sequence).

### Root Cause
The nested `vmap` operations in `transformers/masking_utils.py` are failing when indexing `padding_mask[batch_idx, kv_idx]`. This is likely a bug in transformers 4.54.0's SDPA (Scaled Dot Product Attention) mask creation that was fixed in later versions, but you're constrained by llmcompressor compatibility.

### Recommended Solutions

#### 1. **Use Sequential Calibration Pipeline** (Quick Fix)
The warning explicitly recommends this. Modify your GPTQ recipe/config:

```python
# In your quantization config/recipe
recipe = """
quant_stage:
  quant_modifiers:
    GPTQModifier:
      config_groups:
        group_0:
          weights:
            num_bits: 4
            type: int
            symmetric: true
          targets: ["Linear"]
      # Add this line
      calibration_pipeline: "sequential"  # Instead of "basic"
"""
```

#### 2. **Disable SDPA Mechanism**
Force the model to use the older attention implementation:

```python
import torch
from transformers import AutoModelForCausalLM

# Load model with SDP disabled
model = AutoModelForCausalLM.from_pretrained(
    "Qwen/Qwen2-1.5B",
    torch_dtype=torch.float16,
    attn_implementation="eager"  # or "flash_attention_2" if available
)
```

#### 3. **Upgrade Transformers Strategically**
Try a newer version that fixes this but might still work with llmcompressor:

```bash
pip install transformers==4.56.1 tokenizers==0.21.1
```

This version is newer but still close enough that compatibility may hold. **If llmcompressor complains**, you may need to pin an older compatible version of llmcompressor too.

#### 4. **Modify Calibration Data Collator**
The issue could be how `neuralmagic/calibration` is formatted. Try preprocessing it explicitly:

```python
from datasets import load_dataset
from transformers import AutoTokenizer

tokenizer = AutoTokenizer.from_pretrained("Qwen/Qwen2-1.5B")

def preprocess(examples):
    return tokenizer(examples["text"], padding="max_length", max_length=1024, truncation=True)

calib_dataset = load_dataset("neuralmagic/calibration", split="train[:64]")
calib_dataset = calib_dataset.map(preprocess, batched=True, remove_columns=calib_dataset.column_names)
calib_dataset.set_format(type="torch", columns=["input_ids", "attention_mask"])
```

#### 5. **Downgrade PyTorch (Last Resort)**
The `vmap` behavior changed in recent PyTorch versions. If using PyTorch 2.5+, try:

```bash
pip install torch==2.4.1 torchvision==0.19.1 --index-url https://download.pytorch.org/whl/rocm6.4
```

### Immediate Action
Try **Solution #1 first** as it's the most targeted fix and aligns with the library's own recommendation. If that fails, **Solution #2** is the next best bet since it bypasses the problematic SDPA masking entirely.

Which approach would you like to try? If you share your GPTQ script/config, I can provide more specific code modifications.

**Comment by jiridanek** [comment]
@coderabbitai analyze the newly added comments about testing the resulting image

File: dependencies/odh-notebooks-meta-llmcompressor-deps/pyproject.toml (hunk 1 above)
[the remark this answers is not in the document]
Comment (atheo89) [reply dependencies/odh-notebooks-meta-llmcompressor-deps/pyproject.toml]: I would say to request from AIPCC and also add it on the dependency file as well, as it is the only one that differs from the rest on the images and we have on the llm compressor

**Comment by jiridanek** [comment]
## Working W4A16 GPTQ Test Script

This is the verified working script for W4A16 GPTQ quantization with llmcompressor 0.9.0:

<details>
<summary><b>test_w4a16_fixed.py</b> (click to expand)</summary>

```python
import torch
from datasets import load_dataset
from transformers import AutoModelForCausalLM, AutoTokenizer
from llmcompressor import oneshot
from llmcompressor.modifiers.quantization import GPTQModifier

MODEL_ID = "Qwen/Qwen2-1.5B"

print(f"Loading model: {MODEL_ID}")
model = AutoModelForCausalLM.from_pretrained(
    MODEL_ID, 
    torch_dtype="auto",
    device_map="auto",
)
tokenizer = AutoTokenizer.from_pretrained(MODEL_ID)

# Calibration dataset
DATASET_ID = "neuralmagic/calibration"
NUM_CALIBRATION_SAMPLES = 10
MAX_SEQUENCE_LENGTH = 512

print(f"Loading calibration dataset: {DATASET_ID}")
ds = load_dataset(DATASET_ID, name="LLM", split=f"train[:{NUM_CALIBRATION_SAMPLES}]")

def preprocess_function(example):
    messages = example["messages"]
    text = "\n".join([m["content"] for m in messages])
    return tokenizer(text, truncation=True, max_length=MAX_SEQUENCE_LENGTH)

print("Preprocessing calibration data...")
ds = ds.map(preprocess_function, batched=False, remove_columns=ds.column_names)

recipe = GPTQModifier(targets="Linear", scheme="W4A16", ignore=["lm_head"])

print("Applying W4A16 GPTQ quantization (WITHOUT data_collator)...")
oneshot(
    model=model,
    dataset=ds,
    recipe=recipe,
    max_seq_length=MAX_SEQUENCE_LENGTH,
    num_calibration_samples=NUM_CALIBRATION_SAMPLES,
)

print("\n========== SAMPLE GENERATION ==============")
input_ids = tokenizer("Hello my name is", return_tensors="pt").input_ids.to(model.device)
output = model.generate(input_ids, max_new_tokens=20, pad_token_id=tokenizer.eos_token_id)
print(tokenizer.decode(output[0]))
print("==========================================")

SAVE_DIR = MODEL_ID.rstrip("/").split("/")[-1] + "-W4A16-fixed"
print(f"\nSaving quantized model to: {SAVE_DIR}")
model.save_pretrained(SAVE_DIR, save_compressed=True)
tokenizer.save_pretrained(SAVE_DIR)

print("\nSUCCESS: W4A16 GPTQ quantization completed without data_collator!")
```

</details>

### Key Points

1. **No custom `data_collator`** - let llmcompressor use its default
2. **Preprocessing** - tokenize with truncation, remove original columns
3. **Recipe** - use `GPTQModifier` with `scheme="W4A16"`
4. **Save** - use `save_compressed=True` for compressed format

**Comment by jiridanek** [comment]
## Analysis: Bug vs. Workaround

### Is There a Bug?

**Yes, there IS a bug** in how transformers' vmap-based attention masking handles certain tensor shapes. However, **it's not a complete blocker** - it only triggers under specific conditions.

### What's the Actual Bug?

The bug is in `transformers/masking_utils.py`:

```python
# In sdpa_mask_recent_torch():
causal_mask = _vmap_for_bhqkv(mask_function)(batch_arange, head_arange, cache_position, kv_arange)
# This expects padding_mask to be 2D (batch, sequence)
# But certain data_collators produce 1D tensors, causing IndexError
```

**The vmap code assumes a specific tensor shape** for `padding_mask`. When a custom data_collator produces tensors with different shapes (e.g., 1D instead of 2D), it crashes.

### Fix vs. Workaround?

| Aspect | Our Solution |
|--------|--------------|
| **Type** | ⚠️ **Workaround**, not a fix |
| **What we did** | Removed custom `data_collator`, use llmcompressor's default |
| **Why it works** | The default collator produces correctly shaped 2D tensors |
| **Root cause fixed?** | No - the vmap bug in transformers still exists |

### The Full Picture

```
Custom data_collator → produces 1D padding_mask → vmap crashes → IndexError
Default data_collator → produces 2D padding_mask → vmap works → Success
```

### Who Should Fix What?

| Component | Issue | Who Should Fix |
|-----------|-------|----------------|
| **transformers** | vmap code doesn't handle 1D padding_mask | HuggingFace |
| **llmcompressor** | Should validate/convert data_collator output shapes | Neural Magic |
| **Documentation** | Should warn about data_collator tensor shape requirements | Both |

### Implications for This PR

The **llmcompressor 0.9.0 upgrade is safe** because:
1. Standard usage (without custom data_collator) works correctly
2. FP8 Dynamic quantization works (no calibration data needed)
3. W4A16 GPTQ works with the default data_collator

However, users who try to use custom data_collators (like in the "well-lit path" docs with Llama-4) may hit this bug if their collator produces 1D tensors.

### Recommendation

1. **Merge the PR** - the upgrade works for standard use cases
2. **Keep [RHOAIENG-45138](https://issues.redhat.com/browse/RHOAIENG-45138) open** - document that custom data_collators need to produce 2D padding_mask tensors
3. **File upstream issues** (if not already) on:
   - [huggingface/transformers](https://github.com/huggingface/transformers) - vmap should handle 1D padding_mask
   - [vllm-project/llm-compressor](https://github.com/vllm-project/llm-compressor) - validate data_collator output shapes

**Review by ide-developer — APPROVED** [review]
